scipy/special/xsf/tools.h

```diff
@@ -261,10 +258,271 @@ namespace detail {
+    template <typename... Args>
+    XSF_HOST_DEVICE inline std::tuple<double, double, int> bracket_root(
+        double (*func)(double, std::tuple<Args...>), double x_left, double x_right, double x_min, double x_max,
+        double factor, bool increasing, std::uint64_t maxiter, std::tuple<Args...> args
+    ) {
+        double y_left = func(x_left, args), y_right = func(x_right, args);
+        double y_left_sgn = std::signbit(y_left), y_right_sgn = std::signbit(y_right);
+
+        if (y_left_sgn != y_right_sgn || (y_left == 0 || y_right == 0)) {
+            /* Check if the initial bracket is valid. */
+            std::tuple<double, double, int> result(x_left, x_right, 0);
+            return result;
+        }
+        bool search_left;
+        /* The frontier is the new leading endpoint of the expanding bracket. The
+         * interior endpoint trails behind the frontier. In each step, except in
+         * a special circumstance, the old frontier endpoint becomes the new interior
+         * endpoint. */
+        double interior, frontier, y_interior, y_frontier, boundary;
+        if ((increasing && y_right < 0) || (!increasing && y_right > 0)) {
+            /* If func is increasing  and func(x_right) < 0 or if func is decreasing and
+             *  f(y_right) > 0, we should expand the bracket to the right. */
+            interior = x_left, frontier = x_right, y_interior = y_left, y_frontier = y_right;
+            search_left = false;
+            boundary = x_max;
+        } else {
+            /* Otherwise we move and expand the bracket to the left. */
+            interior = x_right, frontier = x_left, y_interior = y_right, y_frontier = y_left;
+            search_left = true;
+            boundary = x_min;
+        }
+        bool y_interior_sgn = std::signbit(y_interior);
+        bool y_frontier_sgn = std::signbit(y_frontier);
+
+        /* If f(x_left) = f(x_right), we say we are in a plateau. In this case, only
+         * move the frontier endpoint while keeping the interior endpoint the same (in
+         * contrast to setting the interior endpoint to the old frontier). This helps
+         * avoid moving into a new plateau. */
+        bool plateau = (y_left == y_right);
+        bool stop = false;
+        for (std::uint64_t i = 0; i < maxiter; i++) {
+            double step = (frontier - interior) * factor;
+            if (!plateau) {
+                interior = frontier;
+                y_interior = y_frontier;
+                y_interior_sgn = y_frontier_sgn;
+            }
+            frontier += step;
+            if ((search_left && frontier <= boundary) || (!search_left && frontier >= boundary)) {
+                /* If the frontier has reached the boundary, set it to the boundary and signal
+                 * the algorithm to stop. We cannot search further. */
+                frontier = boundary;
+                stop = true;
+            }
+            y_frontier = func(frontier, args);
```

Comment:
Similar to the above, I think it could help the caller catch a bug earlier by checking for `nan`.